From: Ilija Tovilo <ilija.tovilo@me.com>
Date: Sat, 2 May 2026 01:49:34 +0200
Subject: PHP 7.1-7.2: backport OpenSSL 4 compatibility

Backport the ASN1_STRING accessor and error-buffer changes from
https://github.com/php/php-src/commit/19255f4b784cedf6dd4bb21e97def348d335dd69

Backport notes:
Add OpenSSL 4 SSLv3 guards and retain the ASN1 accessor fallback for
older OpenSSL and LibreSSL versions.
Replace removed fixed-version TLS methods with generic methods and
explicit minimum and maximum protocol versions.

Origin: https://github.com/shivammathur/php-src-backports/commit/b39d5b8abd8dcd20feb0f734ef061a54f8bbd134
---
 ext/openssl/openssl.c     | 20 ++++++++++----------
 ext/openssl/php_openssl.h |  4 +++-
 ext/openssl/xp_ssl.c      | 39 +++++++++++++++++++++++++++++++++------
 3 files changed, 46 insertions(+), 17 deletions(-)

diff --git a/ext/openssl/openssl.c b/ext/openssl/openssl.c
index 476f2690..ba1ce835 100644
--- a/ext/openssl/openssl.c
+++ b/ext/openssl/openssl.c
@@ -730,9 +730,9 @@ void php_openssl_store_errors()
 	errors = OPENSSL_G(errors);
 
 	do {
-		errors->top = (errors->top + 1) % ERR_NUM_ERRORS;
+		errors->top = (errors->top + 1) % PHP_OPENSSL_ERR_BUFFER_SIZE;
 		if (errors->top == errors->bottom) {
-			errors->bottom = (errors->bottom + 1) % ERR_NUM_ERRORS;
+			errors->bottom = (errors->bottom + 1) % PHP_OPENSSL_ERR_BUFFER_SIZE;
 		}
 		errors->buffer[errors->top] = error_code;
 	} while ((error_code = ERR_get_error()));
@@ -910,7 +910,7 @@ static void php_openssl_add_assoc_name_entry(zval * val, char * key, X509_NAME *
 
 static void php_openssl_add_assoc_asn1_string(zval * val, char * key, ASN1_STRING * str) /* {{{ */
 {
-	add_assoc_stringl(val, key, (char *)str->data, str->length);
+	add_assoc_stringl(val, key, (const char *)ASN1_STRING_get0_data(str), ASN1_STRING_length(str));
 }
 /* }}} */
 
@@ -943,12 +943,12 @@ static time_t php_openssl_asn1_time_to_time_t(ASN1_UTCTIME * timestr) /* {{{ */
 	}
 
 	if (timestr_len < 13 && timestr_len != 11) {
-		php_error_docref(NULL, E_WARNING, "unable to parse time string %s correctly", timestr->data);
+		php_error_docref(NULL, E_WARNING, "unable to parse time string %s correctly", ASN1_STRING_get0_data(timestr));
 		return (time_t)-1;
 	}
 
 	if (ASN1_STRING_type(timestr) == V_ASN1_GENERALIZEDTIME && timestr_len < 15) {
-		php_error_docref(NULL, E_WARNING, "unable to parse time string %s correctly", timestr->data);
+		php_error_docref(NULL, E_WARNING, "unable to parse time string %s correctly", ASN1_STRING_get0_data(timestr));
 		return (time_t)-1;
 	}
 
@@ -1542,7 +1542,7 @@ PHP_MINIT_FUNCTION(openssl)
 	}
 
 	php_stream_xport_register("ssl", php_openssl_ssl_socket_factory);
-#ifndef OPENSSL_NO_SSL3
+#if OPENSSL_VERSION_NUMBER < 0x40000000L && !defined(OPENSSL_NO_SSL3)
 	php_stream_xport_register("sslv3", php_openssl_ssl_socket_factory);
 #endif
 	php_stream_xport_register("tls", php_openssl_ssl_socket_factory);
@@ -1615,7 +1615,7 @@ PHP_MSHUTDOWN_FUNCTION(openssl)
 	php_unregister_url_stream_wrapper("ftps");
 
 	php_stream_xport_unregister("ssl");
-#ifndef OPENSSL_NO_SSL3
+#if OPENSSL_VERSION_NUMBER < 0x40000000L && !defined(OPENSSL_NO_SSL3)
 	php_stream_xport_unregister("sslv3");
 #endif
 	php_stream_xport_unregister("tls");
@@ -2241,8 +2241,8 @@ static int openssl_x509v3_subjectAltName(BIO *bio, X509_EXTENSION *extension)
 	}
 
 	extension_data = X509_EXTENSION_get_data(extension);
-	p = extension_data->data;
-	length = extension_data->length;
+	p = ASN1_STRING_get0_data(extension_data);
+	length = ASN1_STRING_length(extension_data);
 	if (method->it) {
 		names = (GENERAL_NAMES*) (ASN1_item_d2i(NULL, &p, length,
 			ASN1_ITEM_ptr(method->it)));
@@ -5877,7 +5877,7 @@ PHP_FUNCTION(openssl_error_string)
 		RETURN_FALSE;
 	}
 
-	OPENSSL_G(errors)->bottom = (OPENSSL_G(errors)->bottom + 1) % ERR_NUM_ERRORS;
+	OPENSSL_G(errors)->bottom = (OPENSSL_G(errors)->bottom + 1) % PHP_OPENSSL_ERR_BUFFER_SIZE;
 	val = OPENSSL_G(errors)->buffer[OPENSSL_G(errors)->bottom];
 
 	if (val) {
diff --git a/ext/openssl/php_openssl.h b/ext/openssl/php_openssl.h
index e357550c..00c4960b 100644
--- a/ext/openssl/php_openssl.h
+++ b/ext/openssl/php_openssl.h
@@ -48,6 +48,8 @@ extern zend_module_entry openssl_module_entry;
 #endif
 #endif
 
+#define PHP_OPENSSL_ERR_BUFFER_SIZE 16
+
 #define OPENSSL_RAW_DATA 1
 #define OPENSSL_ZERO_PADDING 2
 #define OPENSSL_DONT_ZERO_PAD_KEY 4
@@ -69,7 +71,7 @@ extern zend_module_entry openssl_module_entry;
 #include <openssl/err.h>
 
 struct php_openssl_errors {
-	int buffer[ERR_NUM_ERRORS];
+	int buffer[PHP_OPENSSL_ERR_BUFFER_SIZE];
 	int top;
 	int bottom;
 };
diff --git a/ext/openssl/xp_ssl.c b/ext/openssl/xp_ssl.c
index 4864f90c..94acb022 100644
--- a/ext/openssl/xp_ssl.c
+++ b/ext/openssl/xp_ssl.c
@@ -37,6 +37,10 @@
 #include <openssl/x509v3.h>
 #include <openssl/err.h>
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || (defined(LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER < 0x20700000L)
+#define ASN1_STRING_get0_data ASN1_STRING_data
+#endif
+
 #if OPENSSL_VERSION_NUMBER >= 0x10002000L
 #include <openssl/bn.h>
 #include <openssl/dh.h>
@@ -52,7 +56,7 @@
 #undef X509_EXTENSIONS
 #endif
 
-#ifndef OPENSSL_NO_SSL3
+#if OPENSSL_VERSION_NUMBER < 0x40000000L && !defined(OPENSSL_NO_SSL3)
 #define HAVE_SSL3 1
 #endif
 
@@ -432,12 +436,12 @@ static zend_bool php_openssl_matches_san_list(X509 *peer, const char *subject_na
 			}
 			OPENSSL_free(cert_name);
 		} else if (san->type == GEN_IPADD) {
-			if (san->d.iPAddress->length == 4) {
+			if (ASN1_STRING_length(san->d.iPAddress) == 4) {
 				sprintf(ipbuffer, "%d.%d.%d.%d",
-					san->d.iPAddress->data[0],
-					san->d.iPAddress->data[1],
-					san->d.iPAddress->data[2],
-					san->d.iPAddress->data[3]
+					ASN1_STRING_get0_data(san->d.iPAddress)[0],
+					ASN1_STRING_get0_data(san->d.iPAddress)[1],
+					ASN1_STRING_get0_data(san->d.iPAddress)[2],
+					ASN1_STRING_get0_data(san->d.iPAddress)[3]
 				);
 				if (strcasecmp(subject_name, (const char*)ipbuffer) == 0) {
 					sk_GENERAL_NAME_pop_free(alt_names, GENERAL_NAME_free);
@@ -970,10 +974,18 @@ static const SSL_METHOD *php_openssl_select_crypto_method(zend_long method_value
 		return NULL;
 #endif
 	} else if (method_value == STREAM_CRYPTO_METHOD_TLSv1_0) {
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L
+		return is_client ? TLS_client_method() : TLS_server_method();
+#else
 		return is_client ? TLSv1_client_method() : TLSv1_server_method();
+#endif
 	} else if (method_value == STREAM_CRYPTO_METHOD_TLSv1_1) {
 #ifdef HAVE_TLS11
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L
+		return is_client ? TLS_client_method() : TLS_server_method();
+#else
 		return is_client ? TLSv1_1_client_method() : TLSv1_1_server_method();
+#endif
 #else
 		php_error_docref(NULL, E_WARNING,
 				"TLSv1.1 unavailable in the OpenSSL library against which PHP is linked");
@@ -981,7 +993,11 @@ static const SSL_METHOD *php_openssl_select_crypto_method(zend_long method_value
 #endif
 	} else if (method_value == STREAM_CRYPTO_METHOD_TLSv1_2) {
 #ifdef HAVE_TLS12
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L
+		return is_client ? TLS_client_method() : TLS_server_method();
+#else
 		return is_client ? TLSv1_2_client_method() : TLSv1_2_server_method();
+#endif
 #else
 		php_error_docref(NULL, E_WARNING,
 				"TLSv1.2 unavailable in the OpenSSL library against which PHP is linked");
@@ -1600,6 +1616,17 @@ int php_openssl_setup_crypto(php_stream *stream,
 		return FAILURE;
 	}
 
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L
+	if ((method_flags & (method_flags - 1)) == 0) {
+		int version = method_flags == STREAM_CRYPTO_METHOD_TLSv1_0 ? TLS1_VERSION :
+			(method_flags == STREAM_CRYPTO_METHOD_TLSv1_1 ? TLS1_1_VERSION : TLS1_2_VERSION);
+		if (!SSL_CTX_set_min_proto_version(sslsock->ctx, version) ||
+			!SSL_CTX_set_max_proto_version(sslsock->ctx, version)) {
+			return FAILURE;
+		}
+	}
+#endif
+
 	if (GET_VER_OPT("no_ticket") && zend_is_true(val)) {
 		ssl_ctx_options |= SSL_OP_NO_TICKET;
 	}
